kong/dns/client.lua

```diff
@@ -0,0 +1,783 @@
+  local key = "short:" .. name .. ":" .. (qtype or "all")
+
+  stats_init_name(self.stats, name)
+  stats_increment(self.stats, name, "runs")
+
+  if detect_recursion(resolved_names, key) then
+    stats_increment(self.stats, name, "fail_recur")
+    return nil, "recursion detected for name: " .. name
+  end
+
+  -- quickly lookup with the key "short:<name>:all" or "short:<name>:<qtype>"
+  local answers, err, hit_level = self.cache:get(key)
+  if not answers then
+    log(DEBUG, "quickly cache lookup ", key, " ans:- hlvl:", hit_level or "-")
+
+    answers, err, tries = resolve_names_and_types(self, name, qtype, cache_only,
+                                                  key, tries, resolved_names)
+    if not cache_only and answers then
+      -- If another worker resolved the name between these two `:get`, it can
+      -- work as expected and will not introduce a race condition.
+
+      -- insert via the `:get` callback to prevent inter-process communication
+      self.cache:get(key, nil, function()
+        return answers, nil, answers.ttl
+      end)
+    end
+
+    stats_increment(self.stats, name, answers and "miss" or "fail")
+
+  else
+    log(DEBUG, "quickly cache lookup ", key, " ans:", #answers, " hlv:", hit_level or "-")
+
+    local ctx = ngx.ctx
+    if ctx and ctx.has_timing then
+      req_dyn_hook_run_hooks("timing", "dns:cache_lookup",
+                             (hit_level and hit_level < HIT_L3))
+    end
+
+    stats_increment(self.stats, name, HIT_LEVEL_TO_NAME[hit_level])
+  end
+
+  -- dereference CNAME
+  if qtype ~= TYPE_CNAME and answers and answers[1].type == TYPE_CNAME then
+    stats_increment(self.stats, name, "cname")
+    return resolve_all(self, answers[1].cname, qtype, cache_only, tries, resolved_names)
+  end
+
+  return answers, err, tries
+end
+
+
+function _M:resolve(name, qtype, cache_only, tries)
+  return resolve_all(self, name, qtype, cache_only, tries, {})
+end
+
+
+function _M:resolve_address(name, port, cache_only, tries, resolved_names)
+  resolved_names = resolved_names or {}
+
+  local answers, err, tries = resolve_all(self, name, nil, cache_only, tries, resolved_names)
+  if not answers then
+    return nil, err, tries
+  end
+
+  -- non-nil answers and return_random
+  if answers[1].type == TYPE_SRV then
+    local answer = get_next_weighted_round_robin_answer(answers)
+    port = (answer.port ~= 0 and answer.port) or port
+    return self:resolve_address(answer.target, port, cache_only, tries, resolved_names)
+  end
+
+  return get_next_round_robin_answer(answers).address, port, tries
+end
+
+
+-- compatible with original DNS client library
+-- These APIs will be deprecated if fully replacing the original one.
+local dns_client
+
+function _M.init(opts)
+  opts = opts or {}
+  opts.valid_ttl = opts.validTtl
+  opts.error_ttl = opts.badTtl
+  opts.stale_ttl = opts.staleTtl
+  opts.cache_size = opts.cacheSize
+
+  local client, err = _M.new(opts)
```

Comment:
This part I don't fully understand. If someone gave `{ valid_ttl = 30 }` as an input, the code will set it to `nil`. I suggest:

```suggestion
  if opts then
    opts.valid_ttl = opts.valid_ttl or opts.validTtl
    opts.error_ttl = opts.error_ttl or opts.badTtl
    opts.stale_ttl = opts.stale_ttl or opts.staleTtl
    opts.cache_size = opts.cache_size or opts.cacheSize
  end
  local client, err = _M.new(opts)
```

Reply:
done https://github.com/Kong/kong/pull/12305/commits/7a52269633cb766dc3c74d79af9eda0655724897